[FEAT/#93] Group / mvi 적용

TaewoongR opened the pull request:
- closed #93 

## *📍 Work Description*

- Group Card 생성
- 그룹 화면에 LazyGrid 2열로 Group Card 배치
- mvi 주요 세가지 state, event(intent로 명명), sideEffect 선언
- viewmodel 선언

## *📸 Screenshot*

https://github.com/user-attachments/assets/69cc7091-08eb-4977-b89f-89196a9c3f17

## *📢 To Reviewers*
- 태희님이 적용한 mvi 코드와 [A robust MVI implementation with Jetpack Compose](https://proandroiddev.com/a-robust-mvi-with-jetpack-compose-e08882d2c4ff) 블로그 글을 보며 적용했습니다.
- 현재 그룹들을 로딩하는 event(intent) 와 로딩중,로딩완료, 그룹들 세가지를 포함하는 state를 정의하여 사용했습니다.
  - sideEffect는 현재 사용하지 않고 있습니다.
- state, event(intent), sideEffect 클래스는 전부 @Immutable 를 붙혀서 불변으로 설정했습니다.
- Screen에서는 LaunchedEffect를 사용해 uiState를 키로 하여 그룹들을 로딩하고 업데이트된 상태를 받는 로직을 설정했습니다.

## ⏲️Time

    - 4시간

TaewoongR committed:
refactor/#93 : Screen 파일 패키지로 이동

- mvi 세팅시 구분 짓기 위함

TaewoongR committed:
feat/#93 : GroupCard 컴포넌트 생성

TaewoongR committed:
feat/#93 : Card Data 데이터클래스 추가

TaewoongR committed:
feat/#93 : 나의 그룹 LazyGrid 그룹카드 화면 구성

TaewoongR committed:
feat/#93 : 나의 그룹 화면 디자인 깎기

- 나의 그룹 화면 = 그룹 메인 화면

TaewoongR committed:
feat/#93 : Group State, Event(Intent라 명명), SideEffect 정의

TaewoongR committed:
fix/#93 : core:model 의 Group Model 속성 추가

- 다른곳에서 의존중이라 추가 속성에 기본값 부여

TaewoongR committed:
feat/#93 : viewmodel 선언 및 화면에 적용

TaewoongR committed:
feat/#93 : screen, viewmodel 업데이트

- LaunchedEffect 를 사용하여 uistate를 감지하고 intent 전달
- 로드 완료 이벤트 뷰모델로 전달

TaewoongR committed:
fix/#93 : LaunchedEffect 로직 수정

- 로딩중과 로딩완료 조건문 수정으로 무한루프 방지

TaewoongR committed:
fix/#93 : LaunchedEffect 로직 수정2

- 로딩중과 로딩완료 조건문 수정으로 무한루프 방지

TaewoongR committed:
fix/#93 : 안쓰는 데이터 클래스 삭제

TaewoongR committed:
fix/#93 state, event, state 처리 LaunchedEffect 수정

- state 속성 isVisible 삭제 : groups는 최소한 1가지(나만의 그룹) 이상을 가지므로 isNotEmpty = visible로 해석
- event 에 로딩 완료 후 로딩 종료 함수 추가
- 위의 조건에 맞는 LaunchedEffect 설정

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/component/GroupCard.kt:
```diff
@@ -0,0 +1,64 @@
+	title: String,
+	content: String,
+) {
+	Column(
+		modifier = Modifier
+			.padding(vertical = 10.dp)
+			.wrapContentSize()
+			.clip(RoundedCornerShape(16.dp))
+			.clickable(
+				interactionSource = null,
+				indication = MapisodeRippleAIndication,
+				onClick = onGroupDetailClick,
+			),
```
clip 했는데 background에 16.dp 안주면 터치 범위랑 이 column 외형이랑 일치하지 않는 것 아닌가요?

TaewoongR replied:
네 wrapContentSize한 사이즈 그대로 가져요.
<img src="https://github.com/user-attachments/assets/da8c335b-1ac8-4513-9f4d-74bd879a372a" width=200>
카드를 둥글게 해 놓고 리플효과를 넣으니 카드 모서리 부분이 딱 안맞아서 clip 적용했어요.

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/intent/GroupIntent.kt:
```diff
@@ -0,0 +1,9 @@
+
+import androidx.compose.runtime.Immutable
+
+@Immutable
+sealed class GroupIntent {
+	data object LoadGroups : GroupIntent()
+	data object EndLoadingGroups : GroupIntent()
+}
```
컴파일해서 unstable인지 확인해보셨나요? Immutable을 꼭 붙여야 하는지 모르겠어요

TaewoongR replied:
<img src="https://github.com/user-attachments/assets/9033fb8d-873a-4681-abf0-6eca48767d16" width=300>

이것도 첨해보는데 신기하네요.
compose-compiler 와우...
저 리스트는 태희님처럼 persistentList 사용해야겠어. 클래스에만 Immutable 사용하면 stable 될 줄 알았는데 아니었네요.

수정*
이거 관해서 stable로 변경했습니다.
과정이 짧지 않아서 기술문서로 작성했습니다.

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupScreen.kt:
```diff
@@ -47,10 +51,21 @@ private fun GroupScreen(
 	onGroupJoinClick: () -> Unit,
 	onGroupDetailClick: () -> Unit,
 	onGroupCreationClick: () -> Unit,
+	viewModel: GroupViewModel,
```
이렇게 하면 Screen에 대한 재사용성을 가지지 않습니다. Route에서 UiState를 내려주는 방식이 더 낫지 않나요?

TaewoongR replied:
겨우 재사용성에 대해서 이해했네요. 
단순히 Group Feature 모듈에서 사용하는 것에 국한되지 않고 다른 모듈에서도 쓰일 수 있게끔 재사용성을 갖추자는 거군요!!

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupScreen.kt:
```diff
@@ -47,10 +51,21 @@ private fun GroupScreen(
 ) {
+	val uiState = viewModel.uiState.collectAsStateWithLifecycle()
 	val focusManager = LocalFocusManager.current
 	var isMenuPoppedUp by remember { mutableStateOf(false) }
```
rememberSaveable을 사용하지 않은 이유가 있나요?

TaewoongR replied:
compose 내부 저장 상태까지 생각을 못했네요. 같이 적용하는거 좋아보여요

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupScreen.kt:
```diff
@@ -99,65 +114,25 @@ private fun GroupScreen(
-			horizontalAlignment = Alignment.CenterHorizontally,
+				.padding(it),
+			columns = GridCells.Fixed(2),
+			contentPadding = PaddingValues(
+				start = 30.dp,
+				end = 30.dp,
+			),
```
horizontal = 30.dp 로 할 수 있지 않나요?

TaewoongR replied:
아 그렇네요! 이거 수정할게요

haeti-dev reviewed (approved):
고생하셨습니다! 제가 디스코드 공유하기 채널에서 공유한 coil 3.0의 Preview 기능을 써보셔도 좋을 것 같아요~

TaewoongR committed:
fix/#93 : 코드리뷰 반영

- padding start, end -> horizontal
- immutable 의 프로퍼티 List -> PersistentList
- GroupScreen 재사용성을 위해 인자에 직접 viewmodel 대신 uiState 대입, intent 콜은 호이스팅으로 변경

TaewoongR committed:
fix/#93 : coil 프리뷰 적용

- coil test 라이브러리 추가
- GroupCard 프리뷰 적용

TaewoongR committed:
fix/#93 : GroupState 데이터 클래스 stability 확정

- 내부 프로퍼티의 리스트 제너릭을 Wrapper class로 변환해서 stable 로 만듦

shinythink commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/navigation/GroupNavigation.kt:
```diff
@@ -4,11 +4,11 @@ import androidx.navigation.NavController
-import com.boostcamp.mapisode.mygroup.GroupEditScreen
-import com.boostcamp.mapisode.mygroup.GroupJoinScreen
-import com.boostcamp.mapisode.mygroup.MainGroupRoute
+import com.boostcamp.mapisode.mygroup.screen.GroupCreationScreen
+import com.boostcamp.mapisode.mygroup.screen.GroupDetailScreen
+import com.boostcamp.mapisode.mygroup.screen.GroupEditScreen
+import com.boostcamp.mapisode.mygroup.screen.GroupJoinScreen
+import com.boostcamp.mapisode.mygroup.screen.MainGroupRoute
```
워낙 screen이 많으니 이렇게 관리하는 게 더 깔끔하네요

shinythinking reviewed (approved):
주말 동안 고생 많으셨습니다!
저도 빠르게 적용해서 pr 올리겠습니다!
